content/en/docs/reference/access-authn-authz/certificate-signing-requests.md

```diff
@@ -0,0 +1,207 @@
+resulting certificate (and if appropriate, certificate chain) in the `status.certificate` field.
+
+Once the `status.certificate` field has been populated, the request has been completed and clients can now
+fetch the signed certificate PEM data from the CertificateSigningRequest resource.
+
+In order to reduce the number of old CertificateSigningRequest resources left in a cluster, a garbage collection
+controller runs periodically. Depending on the state of the request, it will automatically delete requests older
+than a specified duration:
+
+* Approved requests: automatically deleted after 1 hour
+* Denied requests: automatically deleted after 1 hour
+* Pending requests: automatically deleted after 1 hour
+
+## Signers
+
+### Signers should define...
+
+All signers should provide information about how they work so that clients can predict what will happen to their CSRs.
+This includes:
+
+1. **Trust distribution**: how trust (CA bundles) are distributed.
+1. **Permitted subjects**: (any? specific subtree?) and behavior when a disallowed subject is requested.
+1. **Permitted x509 extensions**: (IP SANs? DNS SANs? Email SANs? URI SANs? others?) and behavior when a disallowed extension is requested.
+1. **Permitted key usages / extended key usages**: (client only? server only? any? signer-determined? CSR-determined?) and behavior when usages different than the signer-determined usages are specified in the CSR.
+1. **Expiration/cert lifetime**: (fixed by signer? configurable by admin? CSR-determined?) and behavior when an expiration different than the signer-determined expiration is specified in the CSR.
+1. **CA bit allowed/disallowed**: and behavior if a CSR contains a request a for a CA cert when the signer does not permit it.
+1. (**optional**) Information about the meaning of additional CERTIFICATE PEM blocks in status.certificate, if different from the standard behavior of treating the additional certificates as intermediates, and presenting them in TLS handshakes.
+
+### Kubernetes signers
+
+Kubernetes provides the following well-known signers:
+
+1. `kubernetes.io/kube-apiserver-client`: signs certificates that will be honored as client-certs by the kube-apiserver.
+  Never auto-approved by kube-controller-manager.
+    1. Trust distribution: signed certificates must be honored as client-certificates by the kube-apiserver. The CA bundle is not distributed by any other means.
+    1. Permitted subjects - no subject restrictions, but approvers and signers may choose not to approve or sign. Certain subjects like cluster-admin level users or groups vary between distributions and installations, but deserve additional scrutiny before approval and signing. An admission plugin is available to restrict system:masters, but it is often not the only cluster-admin subject in a cluster.
+    1. Permitted x509 extensions - Non-critical extensions should be dropped.
```

Comment:
s/should be/are/ (and verify this is actually what happens in the kube-controller-manager signer)

Reply:
👍 this is the relevant policy being applied to all signers: https://github.com/kubernetes/kubernetes/blob/3524a5a32a55acbe7a043896c27ce2019a207dec/pkg/controller/certificates/authority/policies.go#L54-L69

I'm actually not certain if we *do* have this quite right here... This may need a `priority/critical-urgent` bug opening if so. I'll need to do some reading up around x509 to be certain myself so would appreciate any additional eyes here cc @deads2k @mikedanese

Reply:
The template is set up here:
https://github.com/kubernetes/kubernetes/blob/3524a5a32a55acbe7a043896c27ce2019a207dec/pkg/controller/certificates/authority/authority.go#L70-L85

And extensions/extraextensions are cleared by the default policy here https://github.com/kubernetes/kubernetes/blob/3524a5a32a55acbe7a043896c27ce2019a207dec/pkg/controller/certificates/authority/policies.go#L54-L69

That means only the extensions that are parsed into distinct fields (IPAddresses/Emails/URIs/DNSNames) are preserved.

That policy has been in place since 1.17. We should update this to say "all extensions other than subjectAltName and key usages are dropped" to match reality.

For the behavior of our defined signers:
* `kubernetes.io/kube-apiserver-client`
  * any SANs are currently allowed...
* `kubernetes.io/kube-apiserver-client-kubelet`
  * IsKubeletClientCSR ensures no SANs are set
* `kubernetes.io/kubelet-serving`
  * IsKubeletServingCSR ensures DNSNames or IPAddresses is set and EmailAddresses and URIs is not set
* `kubernetes.io/legacy-unknown`
  * any SANs are allowed